applications/plugins/Compliant/mapping/ContactMultiMapping.h

```diff
@@ -0,0 +1,239 @@
+        J1.resize( (int) this->getToModels()[0]->getSize() * self::Nout, (int) this->getFromModels()[0]->getSize() * self::Nin );
+        J2.resize( (int) this->getToModels()[0]->getSize() * self::Nout, (int) this->getFromModels()[1]->getSize() * self::Nin );
+
+        J1.setZero();
+        J2.setZero();
+
+        assert( !contacts->empty() );
+        assert(pairs->size() == contacts->size());
+		
+        for(unsigned i = 0, activatedIndex=0; i < n; ++i)
+        {
+            if( !mask.empty() && !mask[i] ) continue; // not activated
+
+
+            if( self::Nout==2 )
+            {
+                Eigen::Matrix<real, 3, 1> n = utils::map( (defaulttype::Vec<3, real>)(*contacts)[i].normal );
+                try{
+                    local_frame.template rightCols<2>() = ker( n );
+                }
+                catch( const std::logic_error& ) {
+                    std::cout << "skipping degenerate normal for contact " << i
+                            << ": " << n.transpose() << std::endl;
+                    local_frame.setZero();
+                }
+            }
+            else
+            {
+                // first vector is normal
+                local_frame.col(0) = utils::map( (defaulttype::Vec<3, real>)(*contacts)[i].normal );
+
+                // possibly tangent directions
+                if( self::Nout == 3 ) {
+                    Eigen::Matrix<real, 3, 1> n = local_frame.col(0);
+                    try{
+                      local_frame.template rightCols<2>() = ker( n );
+                    }
+                    catch( const std::logic_error& ) {
+                      std::cout << "skipping degenerate normal for contact " << i
+                                << ": " << n.transpose() << std::endl;
+                      local_frame.setZero();
+                    }
+                }
+            }
+            
+            // make sure we're cool
+            assert( !has_nan(local_frame) );
+
+            // rows
+            for( unsigned k = 0; k < self::Nout; ++k) 
+            {
+                unsigned row = self::Nout * activatedIndex + k;
+                J1.startVec( row );
+                J2.startVec( row );
+
+                for( unsigned j = 0; j < self::Nin; ++j) {
+                    real w = local_frame(j, k);
+                    if(w)
+                    {
+                        J1.insertBack(row, (*pairs)[i][0] * Nin + j ) = -w;
+                        J2.insertBack(row, (*pairs)[i][1] * Nin + j ) = w;
+                    }
+                }
+            }
+
+            ++activatedIndex;
+        }
+		J1.finalize();
+        J2.finalize();
+	}
+
+};
+
+
+
+#if defined(SOFA_EXTERN_TEMPLATE) && !defined(SOFA_COMPONENT_MAPPING_CONTACTMULTIMAPPING_CPP)
+#ifndef SOFA_FLOAT
+extern template class SOFA_Compliant_API ContactMultiMapping< defaulttype::Vec3dTypes, defaulttype::Vec1dTypes >;
```

Comment:
Use `defaulttype::Vec3Types` instead of  `defaulttype::Vec3dTypes` (same with Vec1d, Vec2d, Vec1f, Vec2f, ...)